dashboard/modules/job/tests/test_backwards_compatibility.py

```diff
@@ -0,0 +1,67 @@
+import logging
+
+import pytest
+import sys
+import os
+import subprocess
+import uuid
+from contextlib import contextmanager
+
+logger = logging.getLogger(__name__)
+
+
+@contextmanager
+def conda_env(env_name):
+    # Set env name for shell script
+    os.environ["JOB_COMPATIBILITY_TEST_TEMP_ENV"] = env_name
+    # Delete conda env if it already exists
+    try:
+        yield
+    finally:
+        # Clean up created conda env upon test exit to prevent leaking
+        del os.environ["JOB_COMPATIBILITY_TEST_TEMP_ENV"]
+        subprocess.run(
+            f"conda env remove -y --name {env_name}",
+            shell=True,
+            stdout=subprocess.PIPE)
+
+
+def _compatibility_script_path(file_name: str) -> str:
+    return os.path.join(
+        os.path.dirname(__file__), "backwards_compatibility_scripts",
+        file_name)
+
+
+class TestBackwardsCompatibility:
+    def test_cli(self):
+        """
+        1) Create a new conda environment with ray version X installed
+            inherits same env as current conda envionment except ray version
+        2) Start head node and dashboard with ray version X
+        3) Use current commit's CLI code to do sample job submission flow
+        4) Deactivate the new conda environment and back to original place
+        """
+        # Shell script creates and cleans up tmp conda environment regardless
+        # of the outcome
+        env_name = f"jobs-backwards-compatibility-{uuid.uuid4().hex}"
+        with conda_env(env_name):
+            shell_cmd = f"{_compatibility_script_path('test_backwards_compatibility.sh')}"  # noqa: E501
+            process = subprocess.Popen(
+                shell_cmd,
+                shell=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT)
+            while process.poll() is None:
+                print(
+                    process.stdout.readline().decode("utf-8"),
+                    end="")  # This blocks until it receives a newline.
+            # When the subprocess terminates there might be unconsumed output
+            # that still needs to be processed.
+            print(process.stdout.read().decode("utf-8"), end="")
+            process.wait()
+
+            assert process.returncode == 0
```

Comment:
btw, you can replace all of this with `subprocess.check_output`, it does exactly this logic.

Reply:
Also, any reason why we don't just have the script logic in python instead of a big bash script?

Reply:
ah you merged faster than i did :D. 

I originally tried to do it in python as well but then realized conda didn't have much python API support and cli seems to be the most common way. conda also complains quite a bit if i try to run server in `subprocess.run()` that i realized didn't actually switch and restore conda the way i want, and using `os.system()` lead to other errors like ensure running `conda init`. 

So i just ended up doing everything in one bash script like https://sourcegraph.com/github.com/ray-project/ray/-/blob/release/util/pip_download_test.sh

Reply:
> btw, you can replace all of this with `subprocess.check_output`, it does exactly this logic.

hmm wait, i need return code here since bash script returns 1 in error, and subprocess.check_output only gives stdout right ?

Reply:
subprocess checkout_output raises an exception if you get nonzero

Reply:
https://docs.python.org/3/library/subprocess.html#subprocess.check_output